bika/lims/browser/publish/emailview.py

```diff
@@ -430,22 +430,24 @@ def write_sendlog(self):
     def publish_samples(self):
         """Publish all samples of the reports
         """
-        reports = self.reports
-        for report in reports:
-            # publish the primary sample
-            primary_sample = report.getAnalysisRequest()
-            self.publish(primary_sample)
-            # publish the contained samples
-            contained_samples = report.getContainedAnalysisRequests()
-            for sample in contained_samples:
-                # skip the primary sample
-                if sample == primary_sample:
-                    continue
-                self.publish(sample)
+        samples = set()
+
+        # collect primary + contained samples of the reports
+        for report in self.reports:
+            samples.add(report.getAnalysisRequest())
+            samples.update(report.getContainedAnalysisRequests())
+
+        # publish all samples + their partitions
+        for sample in samples:
+            self.publish(sample)
 
     def publish(self, sample):
         """Set status to prepublished/published/republished
         """
+        # publish partitions
+        for partition in sample.getDescendants():
+            self.publish(partition)
```

Comment:
To keep the code consistent regarding transitions, I think it would be better to add a function `def after_prepublish` in `workflow.analysisrequest.events` instead of publishing the descendants one by one here. This `after_prepublish` transition might look similar to [`after_publish` funciton](https://github.com/senaite/senaite.core/blob/0606d79897d2f56bcf59751060b698c239a5b02d/bika/lims/workflow/analysisrequest/events.py#L96-L103):

```python
def after_prepublish(self):
    do_action_to_descendants(analysis_request, "publish")
```